File: backend/app/api/chained_calls.py
# Built with Spec4 AI - https://spec4.ai
"""Thin FastAPI handlers for the chained_calls_example_app, mirroring api/single_call.py.

A chain whose second call failed is returned as **200** with
`status: "critique_failed"`, not as an error. Call 1 succeeded and its output is
in the body -- which is exactly what the capability's escalation path requires
be shown -- and a 4xx/5xx would push the client's error branch and discard it.
Only a failure that leaves *nothing* to show (call 1 itself, or a cap that
cannot cover the chain) is an error status.
"""

import structlog
from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession

from backend.app.chained_calls import service
from backend.app.chained_calls.pipeline import (
    CHAIN_LENGTH,
    ROLE_CRITIC,
    ROLE_WRITER,
    StoryDraft,
)
from backend.app.chained_calls.schemas import (
    ChainedCallsRequest,
    ChainedCallsResponse,
    ChainPlanResponse,
    ChainStep,
    FinalOutput,
    IntermediateOutput,
    QualitySignal,
    RetryCritiqueRequest,
)
from backend.app.chained_calls.service import (
    ChainOutcome,
    GenerationUnavailableError,
    InvalidStoryPromptError,
    UsageLimitReachedError,
)
from backend.app.db.session import get_db_session

logger = structlog.get_logger()

router = APIRouter()

#: The chain, described before it runs. Served from the same constants the
#: pipeline uses, so the roles a visitor is promised and the roles that actually
#: execute cannot drift apart.
_PLAN = ChainPlanResponse(
    steps=[
        ChainStep(
            position=1,
            role=ROLE_WRITER,
            label="Struggling Writer",
            description=(
                "Takes your story idea and drafts a short story in a self-doubting, "
                "hedging voice. Its output is the only thing the second call sees."
            ),
        ),
        ChainStep(
            position=2,
            role=ROLE_CRITIC,
            label="Harsh Critic",
            description=(
                "Receives that exact draft as its input and critiques it bluntly, "
                "quoting a specific detail from it."
            ),
        ),
    ],
    chain_length=CHAIN_LENGTH,
    length_note=(
        f"This demo runs exactly {CHAIN_LENGTH} chained calls per submission to "
        "conserve a shared free-tier budget. The pattern itself supports chains of "
        "any length -- two is this demo's budget, not the pattern's limit."
    ),
)


@router.get("/api/chained-calls/plan")
async def get_plan() -> JSONResponse:
    """Describe both calls before any of them run.

    Touches no database and calls no model. Exists so the feature's criterion
    that the visitor is told each call's role *before* submitting is served
    from the same constants the pipeline runs on, rather than from frontend
    copy kept in sync by hand.

    Google-style docstring per project convention.

    Returns:
        A 200 JSON response describing each step, the fixed chain length, and
        why the length is fixed.
    """
    return JSONResponse(status_code=200, content=_PLAN.model_dump())


@router.post("/api/chained-calls/generate")
async def generate(
    payload: ChainedCallsRequest,
    session: AsyncSession = Depends(get_db_session),
) -> JSONResponse:
    """Run the two-call chain over a story idea.

    Google-style docstring per project convention.

    Args:
        payload: The visitor's story idea.
        session: An async DB session injected per-request, used for the shared
            usage-limit and request-logging bookkeeping.

    Returns:
        A 200 JSON response carrying both labeled output blocks -- or, when the
        critic call failed, the intermediate block with
        `status: "critique_failed"` and a notice. Otherwise a 422 for an
        unusable story idea, or a 503 carrying a machine-readable `code`:
        `usage_limit_reached` (the cap cannot cover both calls; resets at 00:00
        UTC) or `generation_unavailable` (call 1 failed, so there is no chain).
        Those two are reported separately because they are different operator
        problems.
    """
    try:
        outcome = await service.run_chain(session, story_prompt=payload.story_prompt)
    except InvalidStoryPromptError as exc:
        return _error(422, exc.code, str(exc))
    except (UsageLimitReachedError, GenerationUnavailableError) as exc:
        logger.error("chained_calls_failed", code=exc.code, error=str(exc))
        return _error(503, exc.code, str(exc))

    return JSONResponse(status_code=200, content=_body(outcome).model_dump())


@router.post("/api/chained-calls/retry-critique")
async def retry_critique(
    payload: RetryCritiqueRequest,
    session: AsyncSession = Depends(get_db_session),
) -> JSONResponse:
    """Re-run only the critic call against a story that already exists.

    The capability's mitigation for a failed second call. Regenerating the
    story would spend a unit reproducing something the visitor is already
    reading, and would produce a *different* story -- so the critique they
    finally get would not be a critique of the draft on their screen.

    Args:
        payload: The intermediate output from an earlier run.
        session: An async DB session for usage/logging bookkeeping.

    Returns:
        A 200 JSON response in the same shape as `generate`. Otherwise a 422
        for a blank or over-long story, or a 503 with `usage_limit_reached`.
    """
    story = payload.intermediate_output.text.strip()
    if not story:
        return _error(
            422, "invalid_intermediate_output", "There is no story text to critique."
        )
    if len(story) > service.MAX_STORY_CHARS:
        return _error(
            422,
            "invalid_intermediate_output",
            f"That story is {len(story)} characters -- the limit is "
            f"{service.MAX_STORY_CHARS}.",
        )

    draft = StoryDraft(title=payload.intermediate_output.title, story=story)

    try:
        outcome = await service.run_critique_only(session, draft=draft)
    except UsageLimitReachedError as exc:
        logger.error("chained_calls_failed", code=exc.code, error=str(exc))
        return _error(503, exc.code, str(exc))

    return JSONResponse(status_code=200, content=_body(outcome).model_dump())


def _body(outcome: ChainOutcome) -> ChainedCallsResponse:
    """Map a service outcome onto the wire response.

    Args:
        outcome: The chain's result, complete or partial.

    Returns:
        The response body, with each block's `role` stamped by its schema
        default rather than carried up from the model.
    """
    final = None
    if outcome.critique is not None:
        final = FinalOutput(
            text=outcome.critique.critique,
            quoted_detail=outcome.critique.quoted_detail,
        )

    signal = None
    if outcome.signal is not None:
        signal = QualitySignal(**vars(outcome.signal))

    return ChainedCallsResponse(
        status=outcome.status,
        intermediate_output=IntermediateOutput(
            title=outcome.draft.title, text=outcome.draft.story
        ),
        final_output=final,
        quality_signal=signal,
        notice=outcome.notice,
        writer_model=outcome.writer_model,
        critic_model=outcome.critic_model,
    )


def _error(status: int, code: str, detail: str) -> JSONResponse:
    """Build the error body shape every route in this app returns.

    Args:
        status: The HTTP status code.
        code: The machine-readable code the frontend branches on.
        detail: A human-readable explanation.

    Returns:
        A JSONResponse carrying all three.
    """
    return JSONResponse(
        status_code=status, content={"status": "error", "code": code, "detail": detail}
    )

File: frontend/src/api/chainedCalls.ts
// Built with Spec4 AI - https://spec4.ai
/**
 * Typed client for the chained-calls endpoints.
 *
 * Mirrors `singleCall.ts` in shape, including its error messages: `fetch`
 * rejects identically for "the backend isn't running" and "the browser blocked
 * the response for CORS", so the unreachable message names both rather than
 * saying something unactionable.
 */

/** How a run ended. `critique_failed` arrives as a 200 — call 1's output is in the body. */
export type ChainStatus = 'complete' | 'critique_failed'

/** Call 1's block. `role` is stamped by the server from the prompt it sent. */
export interface IntermediateOutput {
  role: 'struggling_writer'
  title: string
  text: string
}

/** Call 2's block. */
export interface FinalOutput {
  role: 'harsh_critic'
  text: string
  /**
   * The phrase the critic took from the story. Rendered as the visible evidence
   * that call 2 actually read call 1's output rather than writing generic
   * commentary that would fit any story.
   */
  quoted_detail: string
}

/**
 * The backend's automated check on that quoted detail.
 *
 * A signal, not a verdict: it establishes the detail is *present* in the story,
 * and establishes nothing about whether the critique is apt. The UI must say
 * which of the two it is — this project's recurring defect is a surface
 * asserting something nothing verified.
 */
export interface QualitySignal {
  quoted_detail_found: boolean
  match_ratio: number
  references_story: boolean
}

/** One completed run of the chain, complete or partial. */
export interface ChainResult {
  status: ChainStatus
  intermediate_output: IntermediateOutput
  final_output: FinalOutput | null
  quality_signal: QualitySignal | null
  /** Plain-language explanation of a partial result. */
  notice: string | null
  /** The slug that served call 1, read off its response. */
  writer_model: string
  critic_model: string | null
}

/** One declared step of the chain, described before it runs. */
export interface ChainStep {
  position: number
  role: string
  label: string
  description: string
}

/** Response body of GET /api/chained-calls/plan. */
export interface ChainPlan {
  steps: ChainStep[]
  chain_length: number
  length_note: string
}

/**
 * What the screen shows when the plan hasn't loaded (or can't).
 *
 * The feature requires the visitor be told each call's job **before**
 * submitting, and this deployment's backend sleeps after ~15 minutes idle — so
 * a first-time visitor routinely arrives before the API answers. Rendering
 * nothing until it does would fail that criterion at exactly the moment it
 * matters most. The server's plan replaces this as soon as it arrives; both are
 * built from the same fixed role constants, and `chained-calls.test.tsx` pins
 * that these roles match the wire types.
 */
export const FALLBACK_PLAN: ChainPlan = {
  steps: [
    {
      position: 1,
      role: 'struggling_writer',
      label: 'Struggling Writer',
      description:
        'Takes your story idea and drafts a short story in a self-doubting, hedging voice. ' +
        'Its output is the only thing the second call sees.',
    },
    {
      position: 2,
      role: 'harsh_critic',
      label: 'Harsh Critic',
      description:
        'Receives that exact draft as its input and critiques it bluntly, quoting a specific ' +
        'detail from it.',
    },
  ],
  chain_length: 2,
  length_note:
    'This demo runs exactly 2 chained calls per submission to conserve a shared ' +
    'budget. The pattern itself supports chains of any length -- two is this demo’s budget, ' +
    'not the pattern’s limit.',
}

/** Mirrors `service.MAX_STORY_PROMPT_CHARS`, so an over-long idea is caught before a round trip. */
export const MAX_STORY_PROMPT_CHARS = 600

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8000'

/**
 * A failure the backend described, carrying its machine-readable code.
 *
 * The code separates a spent daily budget (`usage_limit_reached`, which resets
 * at 00:00 UTC and cannot be retried into working) from an unreachable provider
 * (`generation_unavailable`, which can), without matching on prose.
 */
export class ChainedCallsRequestError extends Error {
  readonly code: string
  readonly status: number

  constructor(message: string, code: string, status: number, cause?: unknown) {
    super(message, { cause })
    this.name = 'ChainedCallsRequestError'
    this.code = code
    this.status = status
  }
}

/** Pull a message and code out of an error response, tolerating FastAPI's own 422 shape. */
function errorFrom(body: unknown, status: number): ChainedCallsRequestError {
  if (typeof body === 'object' && body !== null) {
    const record = body as { detail?: unknown; code?: unknown }
    const code = typeof record.code === 'string' ? record.code : 'request_failed'

    if (typeof record.detail === 'string') {
      return new ChainedCallsRequestError(record.detail, code, status)
    }
    if (Array.isArray(record.detail)) {
      const first = record.detail[0] as { msg?: string } | undefined
      if (first?.msg) {
        return new ChainedCallsRequestError(
          first.msg.replace(/^Value error,\s*/, ''),
          'invalid_request',
          status,
        )
      }
    }
  }
  return new ChainedCallsRequestError(`The backend returned ${status}.`, 'request_failed', status)
}

function unreachable(cause: unknown): ChainedCallsRequestError {
  return new ChainedCallsRequestError(
    `Could not reach the backend at ${API_BASE_URL}. Either it isn't running, ` +
      `or this page's origin (${window.location.origin}) doesn't match the ` +
      `backend's CORS_ORIGIN setting.`,
    'unreachable',
    0,
    cause,
  )
}

/**
 * Fetch the description of both calls, to show before anything runs.
 *
 * @returns The two steps, the fixed chain length, and why the length is fixed.
 * @throws ChainedCallsRequestError if the request cannot be made, is blocked, or returns non-2xx.
 */
export async function fetchChainPlan(): Promise<ChainPlan> {
  const url = `${API_BASE_URL}/api/chained-calls/plan`

  let response: Response
  try {
    response = await fetch(url)
  } catch (cause) {
    throw unreachable(cause)
  }

  if (!response.ok) {
    throw new ChainedCallsRequestError(
      `The backend returned ${response.status} for ${url}.`,
      'request_failed',
      response.status,
    )
  }

  return (await response.json()) as ChainPlan
}

export interface ChainSubmission {
  storyPrompt: string
}

/**
 * Run the whole chain: write a story, then critique that exact story.
 *
 * @param submission - The visitor's story idea.
 * @returns Both labeled blocks, or the intermediate one with `critique_failed`.
 * @throws ChainedCallsRequestError if the request cannot be made, is blocked, or is rejected.
 */
export async function runChain({ storyPrompt }: ChainSubmission): Promise<ChainResult> {
  return post(`${API_BASE_URL}/api/chained-calls/generate`, { story_prompt: storyPrompt })
}

export interface CritiqueRetrySubmission {
  /** The story from the failed run, sent back unchanged so the critique matches what's on screen. */
  intermediateOutput: IntermediateOutput
}

/**
 * Re-run only call 2, against a story that already exists.
 *
 * The story travels back through the client because the server stores nothing —
 * that is the capability's privacy requirement, and this is its cost. Sending
 * the draft unchanged is also what makes the retry meaningful: regenerating it
 * would produce a *different* story, so the critique that finally arrived would
 * not be a critique of the draft the visitor is reading.
 *
 * @param submission - The intermediate output from the failed run.
 * @returns The same result shape, now hopefully complete.
 * @throws ChainedCallsRequestError if the request cannot be made, is blocked, or is rejected.
 */
export async function retryCritique({
  intermediateOutput,
}: CritiqueRetrySubmission): Promise<ChainResult> {
  return post(`${API_BASE_URL}/api/chained-calls/retry-critique`, {
    intermediate_output: {
      role: intermediateOutput.role,
      title: intermediateOutput.title,
      text: intermediateOutput.text,
    },
  })
}

/** Shared POST plumbing for both endpoints, which return the same body shape. */
async function post(url: string, payload: unknown): Promise<ChainResult> {
  let response: Response
  try {
    response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    })
  } catch (cause) {
    throw unreachable(cause)
  }

  const body: unknown = await response.json().catch(() => null)

  if (!response.ok) {
    throw errorFrom(body, response.status)
  }

  return body as ChainResult
}
